### PR TITLE
[Feature] Add HiViT

### PR DESCRIPTION
**Pull request by sunsmarterjie**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

We want to add HiViT network to mmpretrain

## Modification

We add mmpretrain/models/backbone/HiViT.py and corresponding config files

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpretrain/models/backbones/hivit.py [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+        x = self.norm(x)
+        x = self.reduction(x)
+
+        if is_main_stage:
+            x = x[:, :, 0, 0, :]
+        return x
+
+
+@MODELS.register_module()
+class HiViT(nn.Module):
```
Comment (fangyixiao18) [inline mmpretrain/models/backbones/hivit.py:317]: please inherit from `BaseBackbone` and add `init_cfg` to function `__init__` args, you could refer to code of swin

File: mmpretrain/models/backbones/hivit.py [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+
+    @torch.jit.ignore
+    def no_weight_decay(self):
+        return {'pos_embed'}
+
+    @torch.jit.ignore
+    def no_weight_decay_keywords(self):
+        return {'relative_position_bias_table'}
```
Comment (fangyixiao18) [inline mmpretrain/models/backbones/hivit.py:507]: line 501-507 can be removed, you have set it in configs

File: mmpretrain/models/backbones/hivit.py [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+
+    def __init__(self,
+                 arch='base',
+                 img_size=224,
+                 patch_size=16,
+                 inner_patches=4,
+                 in_chans=3,
+                 embed_dim=512,
+                 depths=[4, 4, 20],
+                 num_heads=8,
```
Comment (fangyixiao18) [inline mmpretrain/models/backbones/hivit.py:388]: are these 3 args redundant, which are defined in arch_zoo

File: mmpretrain/models/backbones/hivit.py [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+                 attn_drop_rate=0.,
+                 drop_path_rate=0.0,
+                 norm_cfg=dict(type='LN'),
+                 ape=True,
+                 rpe=True,
+                 patch_norm=True,
+                 frozen_stages=-1,
+                 kernel_size=None,
+                 pad_size=None,
+                 layer_scale_init_value=0.0):
```
Comment (fangyixiao18) [inline mmpretrain/models/backbones/hivit.py:403]: better to add `out_indices` to control the output from which block

File: mmpretrain/models/backbones/hivit.py [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed.reshape(1, int(math.sqrt(N)), int(math.sqrt(N)), dim).permute(0, 3, 1, 2),
+            scale_factor=(h0 / math.sqrt(N), w0 / math.sqrt(N)),
+            mode='bicubic',
+        )
+        assert int(h0) == patch_pos_embed.shape[-2] and int(w0) == patch_pos_embed.shape[-1]
+        patch_pos_embed = patch_pos_embed.permute(0, 2, 3, 1).view(1, -1, dim)
+        return patch_pos_embed
+
+    def forward_features(self, x, mask=None):
```
Comment (fangyixiao18) [inline mmpretrain/models/backbones/hivit.py:530]: the code of `forward_features` can be moved to forward for simplicity

File: mmpretrain/models/backbones/hivit.py [hunk 6]
```diff
@@ -0,0 +1,609 @@
...
+        x = self.pos_drop(x)
+
+        rpe_index = True if self.rpe else None
+
+        for blk in self.blocks[-self.num_main_blocks:]:
+            x = blk(x, rpe_index, mask)
+
+        return x
+
+    def forward(self, x, mask=None):
```
Comment (fangyixiao18) [inline mmpretrain/models/backbones/hivit.py:549]: no need to add mask here

**Comment by Ezra-Yu** [comment]
please sign the CLA.

File: mmpretrain/models/backbones/hivit.py (hunk 5 above)
Comment (sunsmarterjie) [reply mmpretrain/models/backbones/hivit.py:530]: forward_features is needed for self-supervised pre-training

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @sunsmarterjie,

We'd like to express our appreciation for your valuable contributions to the mmpretrain. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

File: configs/_base_/models/mae_hivit-base-p16.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# model settings
+model = dict(
+    type='MAEHiViT',
```
Comment (fangyixiao18) [inline configs/_base_/models/mae_hivit-base-p16.py:3]:
```suggestion
    type='MAE',
```

File: configs/_base_/models/mim_itpn-base-p16.py
Comment (fangyixiao18) [inline configs/_base_/models/mim_itpn-base-p16.py:1]: the file is better to be renamed to `itpn_hivit-base-p16.py`

File: configs/_base_/models/mim_itpn-base-p16.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
+# model settings
+model = dict(
+    type='iTPN',
+    backbone=dict(type='MIMiTPN', arch='base', reconstruction_type='pixel', mask_ratio=0.75),
```
Comment (fangyixiao18) [inline configs/_base_/models/mim_itpn-base-p16.py:4]:
```suggestion
    backbone=dict(type='iTPNHiViT', arch='base', reconstruction_type='pixel', mask_ratio=0.75),
```

File: mmpretrain/models/selfsup/itpn.py [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+            image (torch.Tensor): The image tensor with shape NCHW.
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]: The feature and attention mask.
+        """
+        return self.encode_image(image)
+
+
+@MODELS.register_module()
+class MIMiTPN(HiViT):
```
Comment (fangyixiao18) [inline mmpretrain/models/selfsup/itpn.py:148]:
```suggestion
class iTPNHiViT(HiViT):
```

File: mmpretrain/models/selfsup/hivit.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+            x = self.pos_drop(x)
+
+            for blk in self.blocks[-self.num_main_blocks:]:
+                x = blk(x)
+
+            return (x, mask, ids_restore)
+
+
+@MODELS.register_module()
+class MAEHiViT(BaseSelfSupervisor):
```
Comment (fangyixiao18) [inline mmpretrain/models/selfsup/hivit.py:190]: this can be removed, which is the same as MAE

File: mmpretrain/models/selfsup/hivit.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+from mmpretrain.models import HiViT
+from ..utils import build_2d_sincos_position_embedding
+from .base import BaseSelfSupervisor
+from mmpretrain.structures import DataSample
+from mmpretrain.registry import MODELS
+
+
+@MODELS.register_module()
+class MIMHiViT(HiViT):
```
Comment (fangyixiao18) [inline mmpretrain/models/selfsup/hivit.py:12]: better to rename to MAEHiViT and move to `mae.py`

File: configs/itpn/selfsup-itpn-clipbsup-base-p16-amp-coslr-300e_in1k.py
Comment (fangyixiao18) [inline configs/itpn/selfsup-itpn-clipbsup-base-p16-amp-coslr-300e_in1k.py:1]: The name of the file can follow this format: `itpn-clipbsup_hivit-base-p16_8xb256-amp-coslr-300e_in1k.py
`

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: configs/hivit/selfsup-mae_hivit-base-p16_6xb512-amp-coslr-1600e_in1k.py
Comment (fangyixiao18) [inline configs/hivit/selfsup-mae_hivit-base-p16_6xb512-amp-coslr-1600e_in1k.py:1]: The file of HiViT for MAE pre-training can be moved to MAE folder, and directly renamed to mae_hivit-xxxx.py
''

File: configs/hivit/sup-hivit-base_16xb64_in1k.py
Comment (fangyixiao18) [inline configs/hivit/sup-hivit-base_16xb64_in1k.py:1]: the file name can follow this format: `hivit-base-p16_16xb64_in1k.py`

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

**Comment by fangyixiao18** [comment]
The updated PR https://github.com/open-mmlab/mmpretrain/pull/1584 will be merged, which has the same modification that you have committed. This PR will be closed. Thanks for the contribution.
